### ISSUE TITLE
Consolidate tests

### ISSUE THREAD

**Issue by edwinb** [idris-lang/Idris-dev#2577]
The tests currently take a long time to run and it would be nice if they were quicker while still testing everything. There is one cheap way to do this:

There's a lot of tests, especially in regxxx, which merely check some code type checks then quit. This involves starting up idris, loading the prelude, then typechecking, then quit and move on to the next test.

Instead, we could combine a lot of these tests into one source file, and save the repeated loading of the prelude.

There are probably other ways the tests could be consolidated - there hasn't really been much thought put into their organisation so it's worth checking to see if any repeat work, or don't really test anything useful.

**Comment by david-christiansen** [comment]
I'll do this after PhD hand-in if nobody else does first :)

**Comment by jmitchell** [comment]
I'd like to take a stab at the testing mini-project a try if there's still interest.

> Rewrite the test suite to use HSpec or similar to provide nice error reporting, categorization and modular integration of various testing tools.

I'm hoping HSpec or some similar tool can speed things up. HSpec may even let us execute tests in parallel.

It's specifically a Haskell testing library, so that probably means interfacing with Idris as a library rather than executable. Does that seem like a reasonable approach? We could keep a few top-level functional tests around to be safe.

**Comment by david-christiansen** [comment]
The test suite can already run in parallel - e.g. `make test_c -j4` runs four at a time.

I think that it might be difficult to achieve a proper test using Idris as a library, but it might be worth a shot. It's a lot less work to start with just combining more tests into single source files so they execute faster.

**Comment by jfdm** [comment]
Just to chime in, one thing to bear in mind is that AFAIK what constitutes 'Idris-as-a-Library' has not been finalised/debated nor how it should be presented 'as-a-library'.  This may affect how we can test it. 

Second, the current test suite also tests interactions with the idris executable, as well as the functionality of the compiler. These 'emergent' tests should be preserved as well. I have always perceived HSpec for code level testing and not executable interactions. I am quite happy to be corrected on this.

Just to say also, I too am not against using things like HSpec, but am a little more concerned about how best it should be used so that we do not lose any tests, emergent or otherwise, on the way.

**Comment by jmitchell** [comment]
All great points. Whitebox testing is a separate concern with its own set of demands. The next obvious step is exactly what @edwinb suggested.

**Comment by jfdm** [comment]
As an addition to this,  PR #2853 introduces the generation of ad hoc timing data for the tests. [A raw dump of the output that has been grep'd is available online](http://jfdm.host.cs.st-andrews.ac.uk/idris/idris-c.dat).

**Comment by steshaw** [comment]
I'd like to start working on this if no-one is current doing so. @david-christiansen? @jmitchell?

**Comment by jfdm** [comment]
@steshaw AFAIK nobody has been working on this. So go ahead. 

I did make an initial attempt at consolidating a few tests and am happy to discuss ideas on how to reduce timings further.

I started by grouping similar idris tests that are expected to pass under single 'well defined' packages.

**Comment by steshaw** [comment]
@jfdm, great because I already started looking at it :).

It seems to help to pass multiple files to the `idris` executable. In this way the files don't need to be into one source file. However, I noticed many tests that use `--check` are expecting the compiler to fail. When this happens, further files are not processed. I was thinking to add a command option ("--continue" ?) to allow all files to be checked/compiled. But first, the files must be processed in the same order as they are passed on the command line!

Is that what you mean when you say "grouping ... tests that are expected to pass"?

**Comment by jfdm** [comment]
> Is that what you mean when you say "grouping ... tests that are expected to pass"? i.e. that they aren't expected to fail with compiler messages?

No, I actually meant those that successfully compile and execute. For example see #3201. 

As for a continue flag. I am not readily opposed to having it, but I think we should be wary of including one. 

We could also approximate it by providing Idris with a REPL script to load each file in turn. For example, take test `reg035`. A suitable input script could be:

```
:load reg035.idr
:load reg035a.idr
:load reg035b.idr
```

and executed as

```sh

idris --nobanner --nocolour < cmds.txt
```

**Comment by steshaw** [comment]
I think I caught you there will a double-negative. Sorry. Looks like we're on the same page. I guess I couldn't find many examples of tests that used `--check` when they were expected to compile because you'd already addressed that with `regression001`.

I'll have a go at the REPL script idea. Thanks!

**Comment by jfdm** [comment]
No worries.

Good luck.

**Comment by steshaw** [comment]
@jfdm the REPL script idea looks good. It is faster and just requires using --quiet otherwise the expected files become much busier.
